java-checks/src/main/java/org/sonar/java/checks/SelectorMethodArgumentCheck.java

```diff
@@ -0,0 +1,136 @@
+import com.google.common.collect.Lists;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.model.declaration.VariableTreeImpl;
+import org.sonar.java.resolve.Symbol.VariableSymbol;
+import org.sonar.java.resolve.Type;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BlockTree;
+import org.sonar.plugins.java.api.tree.ConditionalExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Modifier;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.text.MessageFormat;
+import java.util.Collection;
+import java.util.List;
+
+@Rule(
+  key = "S2301",
+  name = "Public methods should not contain selector arguments",
+  tags = {"design"},
+  priority = Priority.MAJOR)
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.MODULARITY)
+@SqaleConstantRemediation("15min")
+public class SelectorMethodArgumentCheck extends SubscriptionBaseVisitor {
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodTree methodTree = (MethodTree) tree;
+    List<VariableSymbol> booleanParameterSymbols = getBooleanParametersAsSymbol(methodTree.parameters());
+    BlockTree blockTree = methodTree.block();
+
+    if (isPublic(methodTree) && blockTree != null && !booleanParameterSymbols.isEmpty()) {
+      for (VariableSymbol variable : booleanParameterSymbols) {
+        Collection<IdentifierTree> usages = getSemanticModel().getUsages(variable);
+        if (usages.size() == 1) {
+          blockTree.accept(new ConditionnalStatementVisitor(variable.getName(), getSingleUsage(usages), tree));
+        }
+      }
+    }
+  }
+
+  private boolean isPublic(MethodTree methodTree) {
+    return methodTree.modifiers().modifiers().contains(Modifier.PUBLIC);
+  }
+
+  private List<VariableSymbol> getBooleanParametersAsSymbol(List<VariableTree> parameters) {
+    List<VariableSymbol> booleanParameters = Lists.newLinkedList();
+    for (VariableTree variableTree : parameters) {
+      if (isBooleanVariable(variableTree)) {
+        booleanParameters.add(((VariableTreeImpl) variableTree).getSymbol());
+      }
+    }
+    return booleanParameters;
+  }
+
+  private IdentifierTree getSingleUsage(Collection<IdentifierTree> usages) {
+    return Iterables.get(usages, 0);
+  }
+
+  private boolean isBooleanVariable(VariableTree variableTree) {
+    if (!hasSemantic()) {
```

Comment:
There is no point of even executing this rule if we don't have semantic, so this should be at the top of visitNode method.